contrib/platform/src/com/sun/jna/platform/win32/WbemcliUtil.java

```diff
@@ -0,0 +1,700 @@
+            this.nameSpace = nameSpace;
+        }
+
+        /**
+         * @return The class name
+         */
+        public String getWmiClassName() {
+            return wmiClassName;
+        }
+
+        /**
+         * @param wmiClassName
+         *            The classname to set
+         */
+        public void setWmiClassName(String wmiClassName) {
+            this.wmiClassName = wmiClassName;
+        }
+    }
+
+    /**
+     * Helper class wrapping an EnumMap containing the results of a query.
+     */
+    public class WmiResult<T extends Enum<T>> {
+        private Map<T, List<Object>> propertyMap;
+        private Map<T, Integer> vtTypeMap;
+        private int resultCount = 0;
+
+        /**
+         * @param propertyEnum
+         *            The enum associated with this map
+         */
+        public WmiResult(Class<T> propertyEnum) {
+            propertyMap = new EnumMap<T, List<Object>>(propertyEnum);
+            vtTypeMap = new EnumMap<T, Integer>(propertyEnum);
+            for (T type : propertyEnum.getEnumConstants()) {
+                propertyMap.put(type, new ArrayList<Object>());
+                vtTypeMap.put(type, Variant.VT_NULL);
+            }
+        }
+
+        /**
+         * Gets a String value from the WmiResult. This is the return type when
+         * the WMI result is mapped to a BSTR, including results of UINT64 and
+         * DATETIME type which must be further parsed by the user.
+         * 
+         * @param property
+         *            The property (column) to fetch
+         * @param index
+         *            The index (row) to fetch
+         * @return The String containing the specified value, or empty String if
+         *         null
+         */
+        public String getString(T property, int index) {
+            Object o = this.propertyMap.get(property).get(index);
+            if (o == null) {
+                return "";
+            } else if (vtTypeMap.get(property).equals(Variant.VT_BSTR)) {
+                return (String) o;
+            }
+            throw new Wbemcli.WbemcliException(property.name() + " is not a String type.", vtTypeMap.get(property));
+        }
+
+        /**
+         * Gets an Integer value from the WmiResult. This is the return type
+         * when the WMI result is mapped to a VT_I4 (4-byte integer) value,
+         * including results of UINT32 and UINT16. If an unsigned result is
+         * desired, it may require further processing by the user.
+         * 
+         * @param property
+         *            The property (column) to fetch
+         * @param index
+         *            The index (row) to fetch
+         * @return The Integer containing the specified value, or 0 if null
+         */
+        public Integer getInteger(T property, int index) {
+            Object o = this.propertyMap.get(property).get(index);
+            if (o == null) {
+                return 0;
```

Comment:
I'm a bit irritated here - why return a primitive 0 instead of the java `null`? I would understand if you return int, but with a nullable type?

Reply:
That's a very good question, and I went back and forth on this implementation several times.  My reasoning:
- Requiring the user to null-check every return takes away some of the value of a utility class that provides sensible defaults (zero or empty string).  The user still has the ability to test null (and even determine types if not-null) directly using `getValue()`. 
- The return type should be tied directly to the CIM type.  `getInteger()` should only be used for `uint16`, `sint32`, and `uint32` CIM types, all of which suggest a numeric value.  The only time we run into an issue is if WMI has no value for that column, and we don't check the type.
- I've also considered having different return types based on the CIM Type rather than the value type encoded in the Variant.  For example, both `String` and `uint64` return as a Variant BSTR, but the latter requires further parsing (to either `long` or possibly a `BigInteger` or `WinNT.LARGE_INTEGER`.  All of which returns to the question of how to treat parse exceptions / nulls.   I eventually settled on what you see as a happy medium.
- If I had my druthers I'd probably use Java 8's `Optional<Integer>` here.  But that's not an option (pun intended).
- All these numeric values began their life as primitives, from `VARIANT`'s various `.*value()` methods. They only got boxed when being added to the Lists.  When I wrote this, I unboxed and returned primitives, and that's why I initially wrote the defaults.
- In many cases in my application I would immediately re-box them to put into another list or map.  So since the value was already boxed in the list, I felt returning it as it was (with a simple typecast) was the most direct option and changed it.  So the code you see is a relic of the original return type but I still thought the defaults were more useful than null (since the user still has access to test null).
 
From the perspective of a user who cares about null, they could write: